quinn-proto/src/connection/mod.rs

```diff
@@ -479,17 +581,23 @@ where
                 prev.update_unacked = false;
             }
 
-            let mut builder = self.begin_packet(
+            debug_assert!(
+                builder.is_none() && sent_frames.is_none(),
+                "Previous packet must have been finished"
+            );
+
+            builder = Some(self.begin_packet(
                 now,
                 space_id,
                 &mut buf,
                 buf_capacity,
                 (num_datagrams - 1) * (self.path.mtu as usize),
                 ack_eliciting,
-            )?;
+            )?);
+            let builder = builder.as_mut().unwrap();
```

Comment:
Style nit: could we use [`Option::get_or_insert_with`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with) to avoid the unwrap here?

Reply:
Since `begin_packet` is fallible and can't be called inside the closure it would be something along:

```rust
let next_builder = self.begin_packet(
                now,
                space_id,
                &mut buf,
                buf_capacity,
                (num_datagrams - 1) * (self.path.mtu as usize),
                ack_eliciting,
            )?;
            let builder = builder.get_or_insert_with(|| next_builder);
```

Which hides a bit the case that should never happen of "what is if `builder` wasn't None before".

I'm rather indifferent on whether its better.
https://doc.rust-lang.org/std/option/enum.Option.html#method.insert seems to be ideal. But isn't stable yet.

Reply:
Changed now to

```rust
let builder = builder.get_or_insert(self.begin_packet(
                now,
                space_id,
                &mut buf,
                buf_capacity,
                (num_datagrams - 1) * (self.path.mtu as usize),
                ack_eliciting,
            )?);
```